validator/client/src/test/java/tech/pegasys/teku/validator/client/ValidatorRegistratorTest.java

```diff
@@ -0,0 +1,161 @@
+  private final OwnedValidators ownedValidators = mock(OwnedValidators.class);
+  private final ValidatorApiChannel validatorApiChannel = mock(ValidatorApiChannel.class);
+  private final TimeProvider stubTimeProvider = StubTimeProvider.withTimeInMillis(500);
+  private final Signer signer = mock(Signer.class);
+
+  private DataStructureUtil dataStructureUtil;
+
+  private Validator validator1;
+  private Validator validator2;
+  private Validator validator3;
+
+  private ValidatorRegistrator validatorRegistrator;
+
+  @BeforeEach
+  void setUp(SpecContext specContext) {
+    dataStructureUtil = specContext.getDataStructureUtil();
+    validator1 =
+        new Validator(
+            specContext.getDataStructureUtil().randomPublicKey(), signer, Optional::empty);
+    validator2 =
+        new Validator(
+            specContext.getDataStructureUtil().randomPublicKey(), signer, Optional::empty);
+    validator3 =
+        new Validator(
+            specContext.getDataStructureUtil().randomPublicKey(), signer, Optional::empty);
+    validatorRegistrator =
+        new ValidatorRegistrator(
+            forkProvider,
+            ownedValidators,
+            validatorApiChannel,
+            specContext.getSpec(),
+            stubTimeProvider);
+  }
+
+  @TestTemplate
+  void doesNotRegisterValidators_ifNotBeginningOfEpoch() {
+    when(forkProvider.getForkInfo(UInt64.ONE))
+        .thenReturn(SafeFuture.failedFuture(new IllegalStateException("doesn't matter")));
+
+    // initially validators will be registered anyway since it's the first call
+    validatorRegistrator.onSlot(UInt64.ONE);
+    verify(forkProvider).getForkInfo(UInt64.ONE);
+
+    // after the initial call, registration should not occur if not beginning of epoch
+    validatorRegistrator.onSlot(UInt64.ONE);
+    verifyNoMoreInteractions(forkProvider);
+
+    verifyNoInteractions(ownedValidators, validatorApiChannel, signer);
+  }
+
+  @TestTemplate
+  void registersValidators_onBeginningOfEpoch() {
+    // GIVEN
+    ForkInfo forkInfo = dataStructureUtil.randomForkInfo();
+    when(forkProvider.getForkInfo(UInt64.ZERO)).thenReturn(SafeFuture.completedFuture(forkInfo));
+
+    when(ownedValidators.getActiveValidators())
+        .thenReturn(List.of(validator1, validator2, validator3));
+
+    doAnswer(invocation -> SafeFuture.completedFuture(dataStructureUtil.randomSignature()))
+        .when(signer)
+        .signValidatorRegistration(any(ValidatorRegistration.class), eq(UInt64.ZERO), eq(forkInfo));
+
+    // WHEN
+    validatorRegistrator.onSlot(UInt64.ZERO);
+
+    // THEN
+    @SuppressWarnings("unchecked")
+    ArgumentCaptor<SszList<SignedValidatorRegistration>> argumentCaptor =
+        ArgumentCaptor.forClass(SszList.class);
+
+    InOrder inOrder = Mockito.inOrder(validatorApiChannel);
+
+    inOrder.verify(validatorApiChannel, timeout(5000)).registerValidators(argumentCaptor.capture());
+    verifyRegistrations(argumentCaptor.getValue());
```

Comment:
5 seconds in a unit test is a long time...

Reply:
thanks, will keep in mind, changed to 1 second